src/main/java/dev/morling/onebrc/CalculateAverage_jonathanaotearoa.java

```diff
@@ -510,46 +503,65 @@ private int findIndex(int nameHash, final long nameAddress, final byte nameSize)
             // Think about replacing modulo.
             // https://lemire.me/blog/2018/08/20/performance-of-ranged-accesses-into-arrays-modulo-multiply-shift-and-masks/
             int index = (nameHash & 0x7FFFFFFF) % CAPACITY;
-            while (isCollision(index, nameHash, nameAddress, nameSize)) {
+            while (isMismatch(index, nameHash, nameAddress, nameSize)) {
                 index = index == LAST_INDEX ? 0 : index + 1;
             }
             return index;
         }
 
-        private boolean isCollision(final int index, final long nameHash, final long nameAddress, final byte nameSize) {
+        private boolean isMismatch(final int index, final long nameHash, final long nameAddress, final byte nameSize) {
             final StationData existing = table[index];
             if (existing == null) {
                 return false;
             }
-            if (nameHash != existing.nameHash) {
-                return true;
-            }
-            if (nameSize != existing.nameSize) {
-                return true;
+            if (nameHash == existing.nameHash) {
+                if (nameSize == existing.nameSize && meetsEqualityThreshold(nameAddress, existing.nameAddress, nameSize)) {
+                    return false;
+                }
+                // We've got a hash collision :(
+                // final String name1 = loadStringFromMemory(nameAddress, nameSize);
+                // final String name2 = loadStringFromMemory(existing.nameAddress, existing.nameSize);
+                // System.out.printf("Collision. Hash: %d, Name1: '%s', Name2: '%s'%n", nameHash, name1, name2);
             }
-            // Last resort; check if the names are the same.
-            // This is real performance hit :(
-            return !isMemoryEqual(nameAddress, existing.nameAddress, nameSize);
+            return true;
         }
 
         /**
-         * Checks if two locations in memory have the same value.
+         * Checks if the names at the specified locations meet our threshold for equality.
+         * <p>
+         * It's possible, due to hash collisions, that two different names have the same size and hash.
+         * To account for this, we also need to check if the station names themselves are equal.
+         * However, checking all the bytes in both names is costly.
+         * We therefore set a threshold for the maximum number of bytes, at the start and end of the name, to check.
+         * If two names with the same size and hash have the same first N and last N bytes, we're happy they are the equal.
```

Comment:
I'm afraid that this is not valid, as it isn't guaranteed to resolve all potential conflicts.